pennylane/qinfo/classical_fisher.py

```diff
@@ -0,0 +1,174 @@
+
+    def wrapper(*args, **kwargs):
+        loss = functools.partial(circ, **kwargs)
+        if len(args) > 1:
+            return torch.autograd.functional.jacobian(loss, (args))
+        return torch.autograd.functional.jacobian(loss, *args)
+
+    return wrapper
+
+
+def _tf_jac(circ):
+    """TF jacobian as a callable function"""
+    import tensorflow as tf
+
+    def wrapper(*args, **kwargs):
+        with tf.GradientTape() as tape:
+            loss = circ(*args, **kwargs)
+        return tape.jacobian(loss, (args))
+
+    return wrapper
+
+
+def CFIM(qnode, argnums=0):
+    """Computing the classical fisher information matrix (CFIM) using the jacobian of the output probabilities
+    as described in eq. (15) in https://arxiv.org/abs/2103.15191
+    """
+    new_qnode = _make_probs(qnode, post_processing_fn=lambda x: qml.math.squeeze(qml.math.stack(x)))
+
+    interface = qnode.interface
+
+    if interface == "jax":
+        import jax
+
+        jac = jax.jacobian(new_qnode, argnums=argnums)
+
+    if interface == "torch":
+        jac = _torch_jac(new_qnode)
+
+    if interface == "autograd":
+        jac = qml.jacobian(new_qnode)
+
+    if interface == "tf":
+        jac = _tf_jac(new_qnode)
+
+    def wrapper(*args, **kwargs):
+        p = qnode(*args, **kwargs)
+        j = jac(*args, **kwargs)
+
+        # In case multiple variables are used
+        if isinstance(j, tuple):
+            return [_compute_cfim(p, j_i, interface) for j_i in j]
+
+        return _compute_cfim(p, j, interface)
+
+    return wrapper
+
+
+def _compute_cfim(p, dp, interface):
+    r"""Computes the (num_params, num_params) classical fisher information matrix from the probabilities and its derivatives
+    I.e. it computes :math:`CFIM_{ij} = \sum_\ell (\partial_i p_\ell) (\partial_i p_\ell) / p_\ell`
+    """
+
+    # Compute 1/p for p!=0, else 0
+    if any(qml.math.isclose(p, qml.math.zeros_like(p))) and not interface == "tf":
+        mask = p != 0
+        one_over_p = qml.math.zeros_like(p)
+        if interface == "jax":
+            one_over_p = one_over_p.at[mask].set(1 / p[mask])
+        else:
+            one_over_p[mask] = 1 / p[mask]
+    elif interface == "tf":
+        import tensorflow as tf
+        import tensorflow.python.ops.numpy_ops.np_config as np_config
+
+        np_config.enable_numpy_behavior()  # this allows for the manipulations in _compute_cfim with tensors
+        one_over_p = tf.math.divide_no_nan(qml.math.ones_like(p), p)
+    else:
+        one_over_p = 1 / p
+
+    # Multiply dp and p
+    dp_over_p = qml.math.transpose(dp) * one_over_p  # creates (n_params, n_probs) array
+
+    # create final matrix cfim_ij = \sum_l (d_i p_l) (d_j p_l) / p_l
+    if interface == "torch":
+        return dp_over_p @ qml.math.cast(dp, dtype=p.dtype)
+
+    return dp_over_p @ dp  # (n_params, n_probs) @ (n_probs, n_params) = (n_params, n_params)
+
+
+@batch_transform
+def _make_probs(tape, wires=None, post_processing_fn=None):
```

Comment:
Have you thought about accepting only qnodes that return probabilities (`qml.probs`). I think that is something we wanted to avoid ( ignoring the measurement of a QNode). Any thoughts on that @co9olguy ? I think `metric_tensor` was ignoring it.

Reply:
@rmoyard I would argue for 'ignoring' the measurement here, since otherwise it is not possible to train QNodes with expectation values while also computing the fisher information. E.g., workflows where you want to do natural gradient descent (similar to QNG) wouldn't be possible